enhancements/catalog-data-transfer-reduction.md

```diff
@@ -0,0 +1,463 @@
+  - Typically built into an image that can serve packages via gRPC when run to OLM via a CatalogSource
+- *Catalog*: one or more indices serving content.
+
+Catalog invariants:
+- Bundles, channels, and packages have unique names within their parent type (bundle->channel->package->catalog).
+- A package name + bundle name is a unique identifier within a catalog.
+- A channel is invalid if it does not have exactly one head.
+    - This may go away at some point, but can be managed by iterating through each head.
+- A channel is invalid if its replaces or skips references a bundle not in the channel.
+
+It is also worth noting that an operator will fail to install at runtime if a dependency is not met.
+This is not an enforced catalog invariant because a dependency may be satisfied by any available catalog,
+a fact that cannot be known until runtime.
+
+Given the above [goals](#goals), a process that calculates all bundles published between two catalog states is needed.
+The old state is a full catalog or set of packages, channels, and versions (“catalog objects”);
+the new state is a full catalog containing the old state plus newly published catalog objects.
+
+A new `opm diff` command will be introduced, with accompanying libraries, to perform this calculation.
+The output will be a declarative config. There are two modes to `opm diff`:
+- *FromHeads*: diff from input channel heads, and exclude heads from the resulting diff. This can be thought of as a traditional diff.
+- *Expand*: include all input data in the resulting diff. This is effectively a prune operation, where the objects being pruned
+are everything _not_ specified in the input data.
+
+These two modes can be used in conjunction, where a set of catalog objects is expanded first,
+then newer bundles are appended from that diff’s heads to the latest channel state.
+
+Some options that apply to both modes:
+- *Dependencies*: the latest of all possible dependencies of packages in the diff are included (true by default).
+- *LatestHeads*: the newest channel heads for all packages in the new catalog are added to the diff (false by default). This fulfills the discovery goal.
+
+#### Diff algorithm
+
+It is helpful to think about a diff operation in terms of the smallest catalog unit, the bundle,
+because a diff operates on a per-bundle granularity to effectively produce a partial catalog.
+Consider the case where an arbitrary set of bundles is chosen from some initial catalog and made into a new index.
+This case is actually the general case: all diff-like operations described in `opm diff` have such a set:
+- The diff between an existing and updated catalog (FromHeads)
+- The input data itself (Expand)
+- Either of the above two plus Dependencies and/or LatestHeads
+
+Diffing is a complex operation on a catalog, in part because the partial catalog potentially has missing references.
```

Comment:
For what it's worth, a lot of the complexity here isn't actually needed because if you're using declarative config as output, you don't actually need all replacement edge references to exist. The OLM resolver doesn't care about whether a version that's being replaced by another version exists, that is purely a restraint of opm's sqlite layer.

Reply:
Are you suggesting that tombstones are not needed?